documentation/smg/17_pixel_data.md

```diff
@@ -0,0 +1,449 @@
+  the data, and thus, the objects are not independent.
+
+  Using a handle class is also convenient when working with setters.
+  Handle classes allow you to do the following:
+
+  ```matlab
+  >> sqw_obj.data.pix.set_data('signal', zeros(1, pix.num_pixels));
+  ```
+
+  Whereas, using a value class, one must return and assign the class instance
+  for the setter to take effect:
+
+  ```matlab
+  >> sqw_obj.data.pix = sqw_obj.data.pix.set_data('signal', zeros(1, pix.num_pixels));
+  ```
+
+- Data is loaded on access.
+  - No data is loaded from file on construction.
+  - All public data accessors _must_ check that data is loaded before returning data.
+
+  This is advantageous as it prevents performing potentially expensive
+  IO operations when pixels are not used.
+  For example, a user may wish to load an SQW object and plot the image data.
+  This does not require pixels, so time and memory is wasted loading what could
+  be gigabytes of data.
+  Note that this also means it is free to move to a given page of data after
+  constructing the object.
+
+- Pixel data are held in memory as `double`s.
+  This is in contrast to in file (including temporary page files),
+  where pixels are stored as `single`.
+  See [ADR 15](../adr/0015-store-pixel-data-in-single-precision.md).
+
+- `PixelData` objects do not know about their binning.
+  - The binning is defined by the `sqw` object's image.
+
+### Saving Pixel Data
+
+The `PixelData` class has no capacity of its own to save its data to an sqw file.
+This is the responsibility of the sqw file savers;
+in particular, `sqw_binfile_common.put_pix`.
+This method, called when saving an `sqw` object,
+loops over pixel pages and saves them to the `sqw` file.
+
+There also exists `sqw_binfile_common.put_bytes`
+which can be used to save arbitrary data to an `sqw` file.
+You can position the `sqw_binfile_common` instance's file handle at the start
+of the `sqw` file's pixel array,
+and save pixels directory to file this way.
+This is more efficient than saving out to temporary files and then
+reading them back in later when calling `put_pix`.
+
+### Getters/Setters
+
+All getters return data in-memory.
+All setters must set data that is in memory.
+Attempting to get or set data outside the pixel array range will throw an error.
+
+### Properties
+
+There are
+[dependent properties](https://www.mathworks.com/help/matlab/matlab_oop/access-methods-for-dependent-properties.html)
+for each of the nine fields of pixel data.
+There are also dependent properties for common groups of properties,
+e.g. `.coordinates` retrieves `u1`, `u2`, `u3` and `dE`.
+These properties can be used to retrieve and set data _in the current page_.
+
+```matlab
+>> size(pix.signal)
+    ans =
+      [1, pix.num_pixels]
+
+>> head_run_idx = pix.run_idx(1:10);
+>> size(head_run_idx)
+    ans =
+      [1, 10]
+
+>> tail_coordinates = pix.coordinates((end - 10):end);
+>> size(tail_coordinates)
+    ans =
+      [4, 10]
```

Comment:
Not sure this is right? If end==100, then end-10==90 and the range 90:100 has size 11.